analysis/src/vm/function_desugar.rs

```diff
@@ -0,0 +1,207 @@
+        ControlFlow::Continue::<()>(())
+    });
+    e
+}
+
+impl<T: FieldElement> FunctionDesugar<T> {
+    fn generate(&self, file: AnalysisASMFile<T>) -> AnalysisASMFile<T> {
+        AnalysisASMFile {
+            machines: file
+                .machines
+                .into_iter()
+                .map(|(name, m)| (name, self.generate_machine_rom(m)))
+                .collect(),
+        }
+    }
+
+    fn generate_machine_rom(&self, mut machine: Machine<T>) -> Machine<T> {
+        if machine.has_pc() {
+            let latch = "instr_return";
+
+            // the number of inputs is the max of the number of inputs needed in each operation
+            let input_count = machine
+                .functions
+                .iter()
+                .map(|o| o.params.inputs.params.len())
+                .max()
+                .unwrap_or(0);
+            let output_count = machine
+                .functions
+                .iter()
+                .map(|o| {
+                    o.params
+                        .outputs
+                        .as_ref()
+                        .map(|o| o.params.len())
+                        .unwrap_or(0)
+                })
+                .max()
+                .unwrap_or(0);
+
+            // add the return instruction. its exact implementation is responsibility of romgen
+            machine
+                .instructions
+                .push(parse_instruction_definition(&format!(
+                    "instr return {} {{ }}",
+                    (0..output_count)
+                        .map(output_at)
+                        .collect::<Vec<_>>()
+                        .join(", "),
+                )));
+
+            // create one register for each input. They are created after as their reset is different (unconstrained, not 0)
+            let input_assignment_registers_declarations = (0..input_count)
+                .map(|i| parse_register_declaration::<T>(&format!("reg {}[@r];", input_at(i))));
+
+            // create one assignment register for each output
+            let output_assignment_registers_declarations = (0..output_count)
+                .map(|i| parse_register_declaration::<T>(&format!("reg {}[<=];", output_at(i))));
+
+            machine.registers.extend(
+                input_assignment_registers_declarations
+                    .chain(output_assignment_registers_declarations),
+            );
+
+            // replace the parameters in all functions
+            machine.functions = machine
+                .functions
+                .into_iter()
+                .map(|mut f| {
+                    // create substitution map from declared input to the hidden witness columns
+                    let inputs = f
+                        .params
+                        .inputs
+                        .params
+                        .iter()
+                        .enumerate()
+                        .map(|(index, param)| (param.name.clone(), index))
```

Comment:
so this bit and line 96 above rely on the same enumeration but they're not connected. could they be done together somehow, or chained directly? also applicable to line 122 below

Reply:
I don't think so: declaration of the input and output registers is global to the machine, while remapping of function inputs and outputs to these global inputs and outputs happens at the function level.